src/plugins/discord/mirror.js

```diff
@@ -0,0 +1,397 @@
+// @flow
+
+import type Database from "better-sqlite3";
+import stringify from "json-stable-stringify";
+import * as Model from "./models";
+import dedent from "../../util/dedent";
+
+const VERSION = "DISCORD_MIRROR_v1";
+
+/**
+ * Persists a local copy of data from a Discord Guild.
+ * Implements create and read functionality.
+ *
+ * Each Mirror instance is tied to a particular Guild. Trying to use a mirror
+ * for multiple Discord Guilds is not permitted; use separate Mirrors.
+ */
+
+export class Mirror {
+  +_db: Database;
+
+  /**
+   * Construct a new Mirror instance.
+   *
+   * Takes a Database and GuildId.
+   *
+   */
+
+  constructor(db: Database, guildId: Model.Snowflake) {
+    this._db = db;
+    this._transaction(() => {
+      this._initialize(guildId);
+    });
+  }
+
+  _transaction(queries: () => void) {
+    const db = this._db;
+    if (db.inTransaction) {
+      throw new Error("already in transaction");
+    }
+    try {
+      db.prepare("BEGIN").run();
+      queries();
+      if (db.inTransaction) {
+        db.prepare("COMMIT").run();
+      }
+    } finally {
+      if (db.inTransaction) {
+        db.prepare("ROLLBACK").run();
+      }
+    }
+  }
+
+  _initialize(guild: Model.Snowflake) {
+    const db = this._db;
+    // We store the config in a singleton table `meta`, whose unique row
+    // has PRIMARY KEY `0`. Only the first ever insert will succeed; we
+    // are locked into the first config.
+    db.prepare(
+      dedent`\
+        CREATE TABLE IF NOT EXISTS meta (
+            zero INTEGER PRIMARY KEY,
```

Comment:
Should also have the `NOT NULL` primary key.

Reply:
We should open an issue to update this everywhere in the repository?

Reply:
Yes, good idea :] Tracking lessons learned / technical debt in issues I'd support in general.

Reply:
Updated